File: app/src/main/java/net/nashihara/naroureader/views/adapters/NovelDetailRecyclerViewAdapter.java
package net.nashihara.naroureader.views.adapters;

import android.content.Context;
import android.content.res.Resources;
import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.support.annotation.Nullable;
import android.support.v7.util.SortedList;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import net.nashihara.naroureader.R;
import net.nashihara.naroureader.databinding.ItemRankingRecyclerBinding;
import net.nashihara.naroureader.models.entities.NovelItem;

import java.util.List;

import narou4j.entities.Novel;
import narou4j.entities.NovelRank;

public class NovelDetailRecyclerViewAdapter extends RecyclerView.Adapter<NovelDetailRecyclerViewAdapter.BindingHolder> {
    private static final String TAG = NovelDetailRecyclerViewAdapter.class.getSimpleName();

    private LayoutInflater mInflater;
    private SortedList<NovelItem> mSortedList;
    private OnItemClickListener mListener;
    private RecyclerView mRecyclerView;
    private boolean isSearch;

    public NovelDetailRecyclerViewAdapter(Context context, boolean isSearch) {
        this.mInflater = LayoutInflater.from(context);
        this.isSearch = isSearch;
        mSortedList = new SortedList<>(NovelItem.class, new SortedListCallback(this, isSearch));
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        super.onAttachedToRecyclerView(recyclerView);
        mRecyclerView = recyclerView;
    }

    @Override
    public void onDetachedFromRecyclerView(RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        mRecyclerView = null;
    }

    @Override
    public BindingHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        final View v = mInflater.inflate(R.layout.item_ranking_recycler, parent, false);
        return new BindingHolder(v, mListener);
    }

    @Override
    public void onBindViewHolder(BindingHolder holder, int position) {
        if (mSortedList != null && mSortedList.size() > position && mSortedList.get(position) != null) {
            ItemRankingRecyclerBinding binding = holder.getBinding();

            NovelItem novelItem = mSortedList.get(position);
            Novel novel = novelItem.getNovelDetail();

            if (isSearch) {
                setView4Search(binding, novel);
            }
            else {
                NovelRank rank = novelItem.getRank();
                NovelRank prevRank = novelItem.getPrevRank();
                setView4Ranking(binding, novel, rank, prevRank, position);
            }
        }
    }

    @Override
    public int getItemCount() {
        if (mSortedList == null) {
            return 0;
        }
        return mSortedList.size();
    }

    private void setView4Search(ItemRankingRecyclerBinding binding, Novel novel) {
        binding.ranking.setVisibility(View.GONE);
        binding.rankNew.setVisibility(View.GONE);
        binding.rankDiffKigou.setVisibility(View.GONE);
        binding.prevRankText.setVisibility(View.GONE);
        binding.rankNew.setVisibility(View.GONE);
        binding.rankingPoint.setVisibility(View.GONE);

        binding.title.setText(novel.getTitle());
        binding.ncode.setText(novel.getNcode());
        binding.writer.setText(novel.getWriter());
        binding.genre.setText(novel.getGenre().getText());
        binding.allStory.setText(novel.getStory());
        binding.allStory.setVisibility(View.GONE);
        binding.keyword.setText("キーワード：" + novel.getKeyword());
        binding.keyword.setVisibility(View.GONE);
        binding.btnExpand.setAlpha(0.7f);

        if (novel.getIsNovelContinue() == 1) {
            Resources res = mRecyclerView.getResources();
            binding.isContinue.setTextColor(res.getColor(android.support.v7.appcompat.R.color.secondary_text_default_material_light));
            binding.isContinue.setText("連載中");
        } else {
            Resources res = mRecyclerView.getResources();
            binding.isContinue.setTextColor(res.getColor(R.color.colorAccent));
            binding.isContinue.setText("完結済");
        }
        binding.page.setText("全" + String.valueOf(novel.getAllNumberOfNovel()) + "部分");
        binding.length.setText(int2String(novel.getNumberOfChar()) + " 文字");
    }

    private void setView4Ranking(ItemRankingRecyclerBinding binding, Novel novel, NovelRank rank, NovelRank prevRank, int position) {
        if (rank.getRankingType() == null) {
            binding.ranking.setText(String.valueOf(position +1) + "位");
        }
        else {
            binding.ranking.setText(rank.getRank() + "位");
            if (prevRank != null) {
                binding.rankNew.setVisibility(View.GONE);
                binding.rankDiffKigou.setVisibility(View.VISIBLE);
                switch (rank.getRankingType()) {
                    case DAILY: {
                        binding.prevRankText.setText("前日：" + String.valueOf(prevRank.getRank()) + "位");
                        break;
                    }
                    case WEEKLY: {
                        binding.prevRankText.setText("前週：" + String.valueOf(prevRank.getRank()) + "位");
                        break;
                    }
                    case MONTHLY: {
                        binding.prevRankText.setText("前月：" + String.valueOf(prevRank.getRank()) + "位");
                        break;
                    }
                    case QUARTET: {
                        binding.prevRankText.setText("前月：" + String.valueOf(prevRank.getRank()) + "位");
                        break;
                    }
                }
                if (rank.getRank() < prevRank.getRank()) {
//                    binding.prevRankText.setText("前回のランキングから" + String.valueOf(diff) + "位上昇しました！");
                    binding.rankDiffKigou.setImageResource(R.drawable.ic_up);
                }
                else if (rank.getRank() > prevRank.getRank()) {
//                    binding.prevRankText.setText("前回のランキングから" + String.valueOf(diff) + "位下降しました...");
                    binding.rankDiffKigou.setImageResource(R.drawable.ic_down);
                }
                else if (rank.getRank() == prevRank.getRank()) {
//                    binding.prevRankText.setText("前回のランキングと同じだよ");
                    binding.rankDiffKigou.setImageResource(R.drawable.ic_sonomama);
                }
            }
            else {
                binding.prevRankText.setText("前回：ー");
                binding.rankDiffKigou.setVisibility(View.GONE);
                binding.rankNew.setVisibility(View.VISIBLE);
                binding.rankNew.setTextColor(Color.RED);
            }
        }

        binding.title.setText(novel.getTitle());
        binding.rankingPoint.setText(int2String(rank.getPt()) + "pt");
        binding.ncode.setText(novel.getNcode());
        binding.writer.setText(novel.getWriter());
        binding.genre.setText(novel.getGenre().getText());
        binding.allStory.setText(novel.getStory());
        binding.allStory.setVisibility(View.GONE);
        binding.keyword.setText("キーワード：" + novel.getKeyword());
        binding.keyword.setVisibility(View.GONE);
        binding.btnExpand.setAlpha(0.7f);

        if (novel.getIsNovelContinue() == 1) {
            Resources res = mRecyclerView.getResources();
            binding.isContinue.setTextColor(res.getColor(android.support.v7.appcompat.R.color.secondary_text_default_material_light));
            binding.isContinue.setText("連載中");
        } else {
            Resources res = mRecyclerView.getResources();
            binding.isContinue.setTextColor(res.getColor(R.color.colorAccent));
            binding.isContinue.setText("完結済");
        }
        binding.page.setText("全" + String.valueOf(novel.getAllNumberOfNovel()) + "部分");
        binding.length.setText(int2String(novel.getNumberOfChar()) + " 文字");
    }

    private String int2String(int number) {
        StringBuilder builder = new StringBuilder();
        String pageString = String.valueOf(number);
        builder.append(pageString);
        int c = 0;
        for (int i = pageString.length(); i > 0; i--) {
            if (c % 3 == 0 && c != 0) {
                builder.insert(i, ",");
            }
            c++;
        }
        return builder.toString();
    }

    public void addDataOf(List<NovelItem> dataList) {
        mSortedList.addAll(dataList);
    }

    public void removeDataOf(List<NovelItem> dataList) {
        mSortedList.beginBatchedUpdates();
        for (NovelItem item : dataList) {
            mSortedList.remove(item);
        }
        mSortedList.endBatchedUpdates();
    }

    public void clearData() {
        mSortedList.clear();
    }

    public SortedList<NovelItem> getList() {
        return this.mSortedList;
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        this.mListener = listener;
    }

    public interface OnItemClickListener {
        void onItemClick(View view, int position, ItemRankingRecyclerBinding binding);
        void onItemLongClick(View view, int position, ItemRankingRecyclerBinding binding);
    }

    static class BindingHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {
        private final ItemRankingRecyclerBinding binding;
        private OnItemClickListener mListener;

        public BindingHolder(View itemView, OnItemClickListener listener) {
            super(itemView);
            this.mListener = listener;
            binding = DataBindingUtil.bind(itemView);

//            binding.btnStory.setOnClickListener(this);
            binding.btnExpand.setOnClickListener(this);
            binding.getRoot().setOnClickListener(this);
            binding.getRoot().setOnLongClickListener(this);
        }

        public ItemRankingRecyclerBinding getBinding(){
            return this.binding;
        }

        @Override
        public void onClick(View v) {
            if (mListener != null) {
                mListener.onItemClick(v, getLayoutPosition(), binding);
            }
        }

        @Override
        public boolean onLongClick(View v) {
            if (mListener == null) {
                return false;
            }

            mListener.onItemLongClick(v, getLayoutPosition(), binding);
            return true;
        }
    }

    private static class SortedListCallback extends SortedList.Callback<NovelItem> {
        private NovelDetailRecyclerViewAdapter adapter;
        private boolean isSearch;

        public SortedListCallback(@Nullable NovelDetailRecyclerViewAdapter adapter, @Nullable boolean isSearch) {
            this.adapter = adapter;
            this.isSearch = isSearch;
        }

        @Override
        public int compare(NovelItem o1, NovelItem o2) {
            if (o2.getRank().getRank() > o1.getRank().getRank()) {
                return -1;
            }
            if (o1.getRank().getRank() == o2.getRank().getRank()) {
                return 0;
            }
            return 1;
        }

        @Override
        public void onInserted(int position, int count) {
            adapter.notifyItemRangeInserted(position, count);
        }

        @Override
        public void onRemoved(int position, int count) {
            adapter.notifyItemRangeRemoved(position, count);
        }

        @Override
        public void onMoved(int fromPosition, int toPosition) {
            adapter.notifyItemMoved(fromPosition, toPosition);
        }

        @Override
        public void onChanged(int position, int count) {
            adapter.notifyItemChanged(position, count);
        }

        @Override
        public boolean areContentsTheSame(NovelItem oldItem, NovelItem newItem) {
            return oldItem.getRank().getRank() == newItem.getRank().getRank();
        }

        @Override
        public boolean areItemsTheSame(NovelItem item1, NovelItem item2) {
            if (item1.getNovelDetail() == null) {
                return item2.getNovelDetail() == null;
            }
            return item1.getNovelDetail().getNcode() == item2.getNovelDetail().getNcode();
        }
    }
}

File: app/src/main/java/net/nashihara/naroureader/adapters/NovelDetailRecyclerViewAdapter.kt
package net.nashihara.naroureader.adapters

import android.content.Context
import android.content.res.Resources
import android.databinding.DataBindingUtil
import android.graphics.Color
import android.support.v7.util.SortedList
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import net.nashihara.naroureader.R
import net.nashihara.naroureader.databinding.ItemRankingRecyclerBinding
import net.nashihara.naroureader.entities.NovelItem
import net.nashihara.naroureader.models.RankingManager

import narou4j.entities.Novel
import narou4j.entities.NovelRank

class NovelDetailRecyclerViewAdapter(context: Context, private val isSearch: Boolean) : RecyclerView.Adapter<NovelDetailRecyclerViewAdapter.BindingHolder>() {

    private var mInflater = LayoutInflater.from(context)
    val list = SortedList(NovelItem::class.java, SortedListCallback(this, isSearch))
    private var mListener: OnItemClickListener? = null
    private var mRecyclerView: RecyclerView? = null

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        super.onAttachedToRecyclerView(recyclerView)
        mRecyclerView = recyclerView
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView?) {
        super.onDetachedFromRecyclerView(recyclerView)
        mRecyclerView = null
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BindingHolder {
        val v = mInflater.inflate(R.layout.item_ranking_recycler, parent, false)
        return BindingHolder(v, mListener)
    }

    override fun onBindViewHolder(holder: BindingHolder, position: Int) {
        if (list.size() > position && list.get(position) != null) {
            val binding = holder.binding

            val novelItem = list.get(position)
            val novel = novelItem.novelDetail

            if (isSearch) {
                setView4Search(binding, novel)
            } else {
                val rank = novelItem.rank
                val prevRank = novelItem.prevRank
                setView4Ranking(binding, novel, rank, prevRank, position)
            }
        }
    }

    override fun getItemCount(): Int {
        return list.size()
    }

    private fun setView4Search(binding: ItemRankingRecyclerBinding, novel: Novel) {
        binding.ranking.visibility = View.GONE
        binding.rankNew.visibility = View.GONE
        binding.rankDiffKigou.visibility = View.GONE
        binding.prevRankText.visibility = View.GONE
        binding.rankNew.visibility = View.GONE
        binding.rankingPoint.visibility = View.GONE

        binding.title.text = novel.title
        binding.ncode.text = novel.ncode
        binding.writer.text = novel.writer
        binding.genre.text = novel.genre.text
        binding.allStory.text = novel.story
        binding.allStory.visibility = View.GONE
        binding.keyword.text = "キーワード：" + novel.keyword
        binding.keyword.visibility = View.GONE
        binding.btnExpand.alpha = 0.7f

        if (novel.isNovelContinue == 1) {
            val res = mRecyclerView!!.resources
            binding.isContinue.setTextColor(res.getColor(android.support.v7.appcompat.R.color.secondary_text_default_material_light))
            binding.isContinue.text = "連載中"
        } else {
            val res = mRecyclerView!!.resources
            binding.isContinue.setTextColor(res.getColor(R.color.colorAccent))
            binding.isContinue.text = "完結済"
        }
        binding.page.text = "全" + novel.allNumberOfNovel.toString() + "部分"
        binding.length.text = int2String(novel.numberOfChar) + " 文字"
    }

    private fun setView4Ranking(binding: ItemRankingRecyclerBinding, novel: Novel, rank: NovelRank, prevRank: NovelRank?, position: Int) {
        val rankingManager = RankingManager(rank, prevRank)

        binding.ranking.text = rankingManager.buildPositionMessage(position)
        binding.prevRankText.text = rankingManager.buildPrevRankingMessage()

        if (rankingManager.hasPrevNovelRank()) {
            binding.rankNew.visibility = View.GONE
            binding.rankDiffKigou.visibility = View.VISIBLE

            if (rankingManager.isRankUp) {
                binding.rankDiffKigou.setImageResource(R.drawable.ic_up)
            } else if (rankingManager.isRankDown) {
                binding.rankDiffKigou.setImageResource(R.drawable.ic_down)
            } else if (rankingManager.isEqual) {
                binding.rankDiffKigou.setImageResource(R.drawable.ic_sonomama)
            }
        } else {
            binding.rankDiffKigou.visibility = View.GONE
            binding.rankNew.visibility = View.VISIBLE
            binding.rankNew.setTextColor(Color.RED)
        }

        binding.title.text = novel.title
        binding.rankingPoint.text = int2String(rank.pt) + "pt"
        binding.ncode.text = novel.ncode
        binding.writer.text = novel.writer
        binding.genre.text = novel.genre.text
        binding.allStory.text = novel.story
        binding.allStory.visibility = View.GONE
        binding.keyword.text = "キーワード：" + novel.keyword
        binding.keyword.visibility = View.GONE
        binding.btnExpand.alpha = 0.7f

        if (novel.isNovelContinue == 1) {
            val res = mRecyclerView!!.resources
            binding.isContinue.setTextColor(res.getColor(android.support.v7.appcompat.R.color.secondary_text_default_material_light))
            binding.isContinue.text = "連載中"
        } else {
            val res = mRecyclerView!!.resources
            binding.isContinue.setTextColor(res.getColor(R.color.colorAccent))
            binding.isContinue.text = "完結済"
        }
        binding.page.text = "全" + novel.allNumberOfNovel.toString() + "部分"
        binding.length.text = int2String(novel.numberOfChar) + " 文字"
    }

    private fun int2String(number: Int): String {
        val builder = StringBuilder()
        val pageString = number.toString()
        builder.append(pageString)
        var c = 0
        for (i in pageString.length downTo 1) {
            if (c % 3 == 0 && c != 0) {
                builder.insert(i, ",")
            }
            c++
        }
        return builder.toString()
    }

    fun addDataOf(dataList: List<NovelItem>) {
        list!!.addAll(dataList)
    }

    fun removeDataOf(dataList: List<NovelItem>) {
        list!!.beginBatchedUpdates()
        for (item in dataList) {
            list.remove(item)
        }
        list.endBatchedUpdates()
    }

    fun clearData() {
        list!!.clear()
    }

    fun setOnItemClickListener(listener: OnItemClickListener) {
        this.mListener = listener
    }

    interface OnItemClickListener {
        fun onItemClick(view: View, position: Int, binding: ItemRankingRecyclerBinding)
        fun onItemLongClick(view: View, position: Int, binding: ItemRankingRecyclerBinding)
    }

    class BindingHolder(itemView: View, private val mListener: OnItemClickListener?) :
            RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {

        val binding: ItemRankingRecyclerBinding = DataBindingUtil.bind<ItemRankingRecyclerBinding>(itemView)

        init {
            //            binding.btnStory.setOnClickListener(this);
            binding.btnExpand.setOnClickListener(this)
            binding.root.setOnClickListener(this)
            binding.root.setOnLongClickListener(this)
        }

        override fun onClick(v: View) {
            mListener?.onItemClick(v, layoutPosition, binding)
        }

        override fun onLongClick(v: View): Boolean {
            if (mListener == null) {
                return false
            }

            mListener.onItemLongClick(v, layoutPosition, binding)
            return true
        }
    }

    private class SortedListCallback(private val adapter: NovelDetailRecyclerViewAdapter, private val isSearch: Boolean) : SortedList.Callback<NovelItem>() {

        override fun compare(o1: NovelItem, o2: NovelItem): Int {
            if (o2.rank.rank > o1.rank.rank) {
                return -1
            }
            if (o1.rank.rank == o2.rank.rank) {
                return 0
            }
            return 1
        }

        override fun onInserted(position: Int, count: Int) {
            adapter.notifyItemRangeInserted(position, count)
        }

        override fun onRemoved(position: Int, count: Int) {
            adapter.notifyItemRangeRemoved(position, count)
        }

        override fun onMoved(fromPosition: Int, toPosition: Int) {
            adapter.notifyItemMoved(fromPosition, toPosition)
        }

        override fun onChanged(position: Int, count: Int) {
            adapter.notifyItemChanged(position, count)
        }

        override fun areContentsTheSame(oldItem: NovelItem, newItem: NovelItem): Boolean {
            return oldItem.rank.rank == newItem.rank.rank
        }

        override fun areItemsTheSame(item1: NovelItem, item2: NovelItem): Boolean {
            if (item1.novelDetail == null) {
                return item2.novelDetail == null
            }
            return item1.novelDetail.ncode === item2.novelDetail.ncode
        }
    }

    companion object {
        private val TAG = NovelDetailRecyclerViewAdapter::class.java.simpleName
    }
}
